src/software/amazon/ion/impl/IonReaderTextSystemX.java

```diff
@@ -235,26 +236,21 @@ private final void load_scalar_value() throws IOException {
                     break;
                 }
             }
-        }
-        else if (token_type == IonTokenConstsX.TOKEN_HEX) {
-            boolean is_negative = (cs.charAt(0) == '-');
+        } else if (token_type == IonTokenConstsX.TOKEN_HEX) {
+            boolean isNegative = (cs.charAt(0) == '-');
             // prefix = is_negative ? "-0x" : "0x";
-            int pos;
-            if (is_negative) {
-                pos = 1;
-            }
-            else {
-                pos = 0;
+            int pos = isNegative ? 1 : 0;
+            if (cs.length() <= (isNegative ? 3 : 2) || Character.toLowerCase(cs.charAt(pos + 1)) != 'x') {
+                parse_error("Invalid hexadecimal value.");
             }
-            assert((cs.length() > 2) && (cs.charAt(pos) == '0') && (cs.charAt(pos+1) == 'x' || cs.charAt(pos+1) == 'X'));
             cs.deleteCharAt(pos);
             cs.deleteCharAt(pos);
-        }
-        else if (token_type == IonTokenConstsX.TOKEN_BINARY) {
+        } else if (token_type == IonTokenConstsX.TOKEN_BINARY) {
             boolean isNegative = (cs.charAt(0) == '-');
-            int position = isNegative ? 1 : 0;
-            cs.deleteCharAt(position);
-            cs.deleteCharAt(position);
+            int pos = isNegative ? 1 : 0;
+            if (cs.length() <= (isNegative ? 3 : 2)) parse_error("Invalid binary int value");
```

Comment:
I suggest making the error messages for hex/binary consistent (note "int" missing from the hex one--either have it in both or neither, your choice).

Also, shouldn't there be a check for `Character.toLowerCase(cs.charAt(pos + 1)) != 'b'` here (like the hex case)?

I'm also noticing that the hex and binary branches are essentially the same, and could be collapsed into a method taking the token type. It presents an opportunity to delete some code.

Reply:
There wasn't a check before I assumed the original author had a reason to write it into the hex logic and not the binary logic. I can add it to binary.

Reply:
It looks like it was an oversight.